File: packages/verixa-python/verixa/compliance_language.py
"""Compliance-language hardening rules — runtime + test-time validation.

The Verixa positioning relies on specific wording. This module makes those
rules machine-checkable so they cannot drift in user-facing text (READMEs,
demo scripts, dossier templates, error messages).

Rules carried from the START brief and AT-Hack0017-002 architecture lock:

1. "every governed action" — never "every action"
2. "Annex IV-aligned runtime technical dossier" — never "regulator-ready Annex IV dossier"
3. "creates evidence to demonstrate and support" — never "proves"
4. Snapshot-based replay (not bit-exact regeneration)
5. Hedged MI300X claims (no absolute performance guarantees)

Functions:
    check_text(text)        -> list of violations (empty = clean)
    assert_clean(text)      -> raises ComplianceLanguageViolation if dirty
    forbidden_phrases()     -> the canonical forbidden-phrase set

This module has no I/O, no network, no DB. It is pure-function and trivial
to unit-test.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Final

# ---------------------------------------------------------------------------
# Canonical forbidden phrases (case-insensitive, word-boundary aware)
# ---------------------------------------------------------------------------

# Each entry: (regex pattern, human-readable rule, suggested replacement)
_FORBIDDEN: Final[tuple[tuple[str, str, str], ...]] = (
    (
        r"\bevery\s+action\b(?!\s+(is\s+)?governed)",
        "Rule 1: 'every action' is too broad; use 'every governed action'.",
        "every governed action",
    ),
    (
        r"\bregulator[-\s]ready\b",
        "Rule 2: 'regulator-ready' overclaims; use 'Annex IV-aligned' or "
        "'Annex IV-aligned runtime technical dossier'.",
        "Annex IV-aligned",
    ),
    (
        r"\bproves?\b",
        "Rule 3: Verixa does not 'prove' AI behaviour; use 'creates evidence "
        "to demonstrate and support' or 'demonstrates'.",
        "creates evidence to demonstrate and support",
    ),
    (
        r"\bproven\b",
        "Rule 3: Verixa does not 'prove' AI behaviour; use 'demonstrated' "
        "or 'evidenced'.",
        "demonstrated",
    ),
    (
        r"\bbit[-\s]?exact\s+(replay|regeneration)\b",
        "Rule 4: Verixa replay is snapshot-based, not bit-exact; use "
        "'snapshot-based replay'.",
        "snapshot-based replay",
    ),
    (
        r"\b(guaranteed|guarantee[ds]?)\s+(MI300X|throughput|latency|performance)\b",
        "Rule 5: MI300X performance claims must be hedged; remove "
        "'guaranteed'.",
        "(remove or hedge)",
    ),
)


@dataclass(frozen=True, slots=True)
class Violation:
    """A single compliance-language rule violation found in text."""

    rule: str
    matched_text: str
    position: int
    suggestion: str

    def __str__(self) -> str:
        return (
            f"[pos {self.position}] {self.rule} "
            f"(matched: {self.matched_text!r}; suggested: {self.suggestion!r})"
        )


class ComplianceLanguageViolation(ValueError):
    """Raised by assert_clean() when forbidden phrases are detected."""

    def __init__(self, violations: list[Violation]) -> None:
        self.violations = violations
        joined = "\n  - ".join(str(v) for v in violations)
        super().__init__(
            f"Compliance-language violations ({len(violations)}):\n  - {joined}"
        )


def forbidden_phrases() -> tuple[str, ...]:
    """Return the canonical set of forbidden patterns (regex strings).

    Useful for documentation, doc-build linters, and CI checks.
    """
    return tuple(pattern for pattern, _rule, _sugg in _FORBIDDEN)


def check_text(text: str) -> list[Violation]:
    """Scan text for forbidden phrases. Empty list = clean.

    Case-insensitive. Returns all violations in document order.
    """
    if not text:
        return []
    violations: list[Violation] = []
    for pattern, rule, suggestion in _FORBIDDEN:
        for match in re.finditer(pattern, text, flags=re.IGNORECASE):
            violations.append(
                Violation(
                    rule=rule,
                    matched_text=match.group(0),
                    position=match.start(),
                    suggestion=suggestion,
                )
            )
    violations.sort(key=lambda v: v.position)
    return violations


def assert_clean(text: str) -> None:
    """Raise ComplianceLanguageViolation if text contains forbidden phrases."""
    violations = check_text(text)
    if violations:
        raise ComplianceLanguageViolation(violations)

File: packages/verixa-ts/src/compliance-language.ts
/**
 * Compliance-language hardening rules — runtime + test-time validation.
 *
 * The Verixa positioning relies on specific wording. This module makes those
 * rules machine-checkable so they cannot drift in user-facing text (READMEs,
 * demo scripts, dossier templates, error messages).
 *
 * Mirror of `verixa.compliance_language` (Python). Both libraries enforce
 * the same canonical forbidden-phrase set so the hardening is identical
 * across the runtime gateway (Python) and the control plane UI (TypeScript).
 *
 * Rules (locked in AT-Hack0017-002 architecture; see START brief):
 * 1. "every governed action" — never "every action"
 * 2. "Annex IV-aligned runtime technical dossier" — never "regulator-ready"
 * 3. "creates evidence to demonstrate and support" — never "proves"
 * 4. Snapshot-based replay (not bit-exact regeneration)
 * 5. Hedged MI300X claims (no absolute performance guarantees)
 *
 * Invariant on FORBIDDEN_RULES patterns:
 *   Every pattern MUST require at least one literal character to match.
 *   No `*`, `?`, or `{0,N}` quantifiers at the top level. This invariant
 *   keeps `checkText` zero-width-match free; if you add a rule that can
 *   match the empty string you MUST also add a forward-progress guard
 *   (`if (match.index === pattern.lastIndex) pattern.lastIndex += 1`)
 *   AND a test that exercises that branch.
 */

/** A single compliance-language rule violation found in text. */
export interface Violation {
  readonly rule: string;
  readonly matchedText: string;
  readonly position: number;
  readonly suggestion: string;
}

/** Error thrown by {@link assertClean} when forbidden phrases are detected. */
export class ComplianceLanguageViolation extends Error {
  public readonly violations: ReadonlyArray<Violation>;

  constructor(violations: ReadonlyArray<Violation>) {
    const lines = violations.map(violationToString).join('\n  - ');
    super(
      `Compliance-language violations (${violations.length}):\n  - ${lines}`,
    );
    this.name = 'ComplianceLanguageViolation';
    this.violations = violations;
    // Preserve prototype chain through transpilation
    Object.setPrototypeOf(this, ComplianceLanguageViolation.prototype);
  }
}

/** Render a Violation as a single-line string (debugging + error messages). */
export function violationToString(v: Violation): string {
  return (
    `[pos ${v.position}] ${v.rule} ` +
    `(matched: ${JSON.stringify(v.matchedText)}; ` +
    `suggested: ${JSON.stringify(v.suggestion)})`
  );
}

interface ForbiddenRule {
  readonly pattern: RegExp;
  readonly rule: string;
  readonly suggestion: string;
}

// Canonical forbidden-phrase set. Patterns are case-insensitive (`i` flag)
// and use the global flag (`g`) so we can iterate all matches in a string.
// Per the file-level invariant, every pattern requires >= 1 literal char.
const FORBIDDEN_RULES: ReadonlyArray<ForbiddenRule> = [
  {
    pattern: /\bevery\s+action\b(?!\s+(is\s+)?governed)/gi,
    rule: "Rule 1: 'every action' is too broad; use 'every governed action'.",
    suggestion: 'every governed action',
  },
  {
    pattern: /\bregulator[-\s]ready\b/gi,
    rule:
      "Rule 2: 'regulator-ready' overclaims; use 'Annex IV-aligned' or " +
      "'Annex IV-aligned runtime technical dossier'.",
    suggestion: 'Annex IV-aligned',
  },
  {
    pattern: /\bproves?\b/gi,
    rule:
      "Rule 3: Verixa does not 'prove' AI behaviour; use 'creates evidence " +
      "to demonstrate and support' or 'demonstrates'.",
    suggestion: 'creates evidence to demonstrate and support',
  },
  {
    pattern: /\bproven\b/gi,
    rule:
      "Rule 3: Verixa does not 'prove' AI behaviour; use 'demonstrated' " +
      "or 'evidenced'.",
    suggestion: 'demonstrated',
  },
  {
    pattern: /\bbit[-\s]?exact\s+(replay|regeneration)\b/gi,
    rule:
      "Rule 4: Verixa replay is snapshot-based, not bit-exact; use " +
      "'snapshot-based replay'.",
    suggestion: 'snapshot-based replay',
  },
  {
    pattern:
      /\b(guaranteed|guarantee[ds]?)\s+(MI300X|throughput|latency|performance)\b/gi,
    rule:
      "Rule 5: MI300X performance claims must be hedged; remove " +
      "'guaranteed'.",
    suggestion: '(remove or hedge)',
  },
];

/** Return the canonical set of forbidden patterns (regex strings). */
export function forbiddenPhrases(): ReadonlyArray<string> {
  return FORBIDDEN_RULES.map((r) => r.pattern.source);
}

/**
 * Scan text for forbidden phrases. Returns all violations in document order
 * (sorted by `position`). Empty array means clean.
 *
 * Note: the file-level invariant guarantees no zero-width matches, so no
 * forward-progress guard is needed inside the iteration.
 */
export function checkText(text: string): Violation[] {
  if (!text) {
    return [];
  }
  const violations: Violation[] = [];
  for (const { pattern, rule, suggestion } of FORBIDDEN_RULES) {
    // Reset lastIndex because the regex is `g`-flagged and stateful.
    pattern.lastIndex = 0;
    let match: RegExpExecArray | null;
    while ((match = pattern.exec(text)) !== null) {
      violations.push({
        rule,
        matchedText: match[0],
        position: match.index,
        suggestion,
      });
    }
  }
  violations.sort((a, b) => a.position - b.position);
  return violations;
}

/**
 * Throw {@link ComplianceLanguageViolation} if `text` contains forbidden
 * phrases. No-op on clean input.
 */
export function assertClean(text: string): void {
  const violations = checkText(text);
  if (violations.length > 0) {
    throw new ComplianceLanguageViolation(violations);
  }
}
